SR-1188 -- Improve error messages for package definitions

Bouke opened the pull request:
Products referencing unknown modules will now produce an error instead of a warning. A missing module should be considered an error in the product definition.

Examples of the error messages:
```
swift-build: error: the target mylib cannot have the executable myexe as a dependency
swift-build: error: these referenced modules could not be found: lib2
swift-build: error: the package has an unsupported layout, unexpected source file(s) found: /tmp/test.c
swift-build: error: the package has an unsupported layout, multiple source roots found: /tmp/Sources, /tmp/src
swift-build: error: the module at /tmp/Sources/lib does not contain any source files
swift-build: error: the module at /tmp/Sources/lib contains mixed language source files
swift-build: error: multiple modules with the name CSV found; modules should have a unique name, accross dependencies
swift-build: error: the product named test references a module that could not be found: CSV
swift-build: error: the product named test doesn't reference any modules
```

aciidb0mb3r commented on Tests/Functional/MiscellaneousTests.swift:
```diff
@@ -353,8 +353,7 @@ class MiscellaneousTestCase: XCTestCase {
         fixture(name: "Miscellaneous/ProductWithMissingModules") { prefix in
-            XCTAssertBuilds(prefix)
-            XCTAssertFileExists(prefix, ".build/debug/libBar.a")
+            XCTAssertBuildFails(prefix)
```
I wonder if its worth having the test cases check the expected error too

Bouke replied:
I've attempted to create a unit test, but it fails to build. See also https://github.com/Bouke/swift-package-manager/tree/SR-1188-unit-test.

kostiakoval commented:
That's define a nice improvement. Since SwiftPM will not have a `doctor` command we should provide "How to Fix it" information together with error.

Example:
```swift
 .noSources(let path):
    return "the module at \(path) does not contain any source files. Add source files to the module or remove it."
```

I'm not sure about the format of that message, maybe a common format with different color output would make it even better. error - red,  fix - yellow 
Example
 ```
error: the module at \(path) does not contain any source files
fix: Add source files to the module or remove it
```

Bouke commented:
I've added fix-its to the error message. The description will be split on "fix:", when printed it will be on a newline and the "fix:" predicate is printed in yellow. Some examples;

```
error: the package has an unsupported layout, unexpected source file(s) found: /Users/bouke/Developer/tmp/test.c.
fix: move the file(s) inside a module.

error: the package has an unsupported layout, multiple source roots found: /Users/bouke/Developer/tmp/Sources, /Users/bouke/Developer/tmp/src.
fix: remove the extra source roots, or add them to the source root exclude list.

error: the module at /Users/bouke/Developer/tmp/Sources/CSV2 contains mixed language source files.
fix: use only a single language within a module.

error: the product named test doesn't reference any modules.
fix: reference one or more modules from the product.

error: the product named test references a module that could not be found: CSV.
fix: reference only valid modules from the product.

error: multiple modules with the name CSV found.
fix: modules should have a unique name, accross dependencies.
```

kostiakoval commented:
LGTM. 
We could potentially improve it in the future by making it more safe and static typed with  an Error subclass with 2 required properties: `error: String, fix: String`. 
But that's for a next PR

Bouke commented:
I was thinking the same, but that would also mean going over all errors and making the protocol available to all packages. Nevertheless I've implemented a `FixableError`, and moved over the errors inside `PackageLoading`.

aciidb0mb3r commented on Sources/Basic/FixableError.swift:
```diff
@@ -0,0 +1,23 @@
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+public protocol FixableError: CustomStringConvertible {
```
I wonder if `Basic` is the best place to put this...Usually for something to be a candidate of `Basic` I imagine if that can be leveraged by any package though I am not sure if thats the correct way.

Bouke replied:
Ideally it would be put in the package related to either building or cli commands. However that would result in a circular dependency, as lower-level packages need access to this protocol. Any suggestions for another module?

abertelrud replied:
I think Basic is good, since we will want to support fixable errors from a variety of subsystems.  I like having a protocol for this, since (as others have said) we eventually want the fix-it information to be an form that an IDE can mechanically apply.

aciidb0mb3r commented on Sources/PackageLoading/PackageExtensions.swift:
```diff
@@ -1,4 +1,4 @@
-/*
+>>/*
```
This looks unintended typo `>>`

abertelrud commented on Sources/Commands/Error.swift:
```diff
@@ -83,3 +93,11 @@ private func print(error: Any) {
 }
+
+private func print(fix: String) {
+    if ColorWrap.isAllowed(for: .stdErr) {
```
Eventually I want to separate out the coloring from the intent, so that code like this just applies a semantic tag and then the coloring is localized (and possibly customizable).  But that's a future improvement — for the moment, this is appropriate given the rest of the code.  One thing at a time... :)

abertelrud commented:
@swift-ci please test

abertelrud commented:
@swift-ci please test

abertelrud commented:
Looks like the test passed, though this PR didn't update.  Merging.
